Add E2E tests for Service, Revision, Route

mgencur opened the pull request:
* service, revision, route describe and describe with print flags
* route list with print flags
* service create for a duplicate service

<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Fixes #

## Proposed Changes

* increase test coverage via new E2E tests

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note

```

navidshaikh commented on test/e2e/basic_workflow_test.go:
```diff
@@ -132,6 +153,14 @@ func (test *e2eTest) serviceUpdate(t *testing.T, serviceName string, args []stri
+	assert.NilError(t, err)
+
+	expectedName := fmt.Sprintf("service.serving.knative.dev/%s", serviceName)
+	assert.Check(t, strings.Contains(out, expectedName))
```
this should be an exact string match IMO, i..e `assert.Equal(t, out, expectedName)`

navidshaikh replied:
as this is not string manipulation and printing on the client side, but machine readable output

navidshaikh commented on test/e2e/basic_workflow_test.go:
```diff
@@ -154,3 +183,29 @@ func (test *e2eTest) serviceDelete(t *testing.T, serviceName string) {
+	assert.NilError(t, err)
+
+	expectedName := fmt.Sprintf("route.serving.knative.dev/%s", routeName)
+	assert.Check(t, strings.Contains(out, expectedName))
```
same as above about exact name matching

navidshaikh commented on test/e2e/basic_workflow_test.go:
```diff
@@ -93,8 +104,18 @@ func (test *e2eTest) serviceList(t *testing.T, serviceName string) {
+	assert.Check(t, util.ContainsAll(out, serviceName))
+}
+
+func (test *e2eTest) serviceCreateDuplicate(t *testing.T, serviceName string) {
```
IMO, this particular test case doesn't belong to `basic_workflow_test` as the filename suggest, but kind of testing for error handling mechanism, we should probably create another workflow which specifically aims at error handling test cases.

maximilien replied:
If you all decide to move this to a more thorough IT (I am ambivalent at this point), please do the same for suggestion above on delete service.

navidshaikh replied:
should be addressed as part of resolution to https://github.com/knative/client/issues/301

navidshaikh commented on test/e2e/revision_workflow_test.go:
```diff
@@ -33,26 +34,57 @@ func TestRevisionWorkflow(t *testing.T) {
+	assert.NilError(t, err)
+
+	expectedName := fmt.Sprintf("revision.serving.knative.dev/%s", revName)
+	assert.Check(t, strings.Contains(out, expectedName))
```
same comment as above, just check for exact match

maximilien commented on test/e2e/basic_workflow_test.go:
```diff
@@ -38,9 +38,14 @@ func TestBasicWorkflow(t *testing.T) {
 	})
 
+	t.Run("create hello service again and get service already exists error", func(t *testing.T) {
+		test.serviceCreateDuplicate(t, "hello")
```
Nice nice, I wonder if we should also try to delete a service that’s already deleted? I would expect an error in case where service is already deleted (non-existent) and if the service is being deleted. Of course in the end the API should do the right thing and the client should return a “sensible” error.

mgencur replied:
Yeah. Good point. I can add a test for deleting a non-existent service to this PR.

maximilien commented:
/ok-to-test

maximilien commented:
/retest

mgencur commented:
hey! Thanks for the reviews. I have fixed all things except moving the test. I also added one more test that checks deleting a non-existent service. 
Regarding moving the test from the basic workflow, I spoke with @navidshaikh and I will shortly create an issue with a proposal for restructuring the test suite a little bit. When we agree on what to do we can move the test(s) to appropriate locations.

mgencur commented:
@navidshaikh the failure doesn't seem to be related to my changes. Can you advise what to do, pls?

rhuss commented:
/retest

rhuss commented:
I'm not sure about that flake, because I thought it was fixed with #271 . It might be that we have to increase the timeout for the wait (as if there's a timeout, then the previous update command returns prematurely).

I just triggered a retest. Let's see how it goes ...

mgencur commented:
Passed this time:)

maximilien commented:
@mgencur not quite...

```bash
Running 'kubectl delete namespace kne2etests0'...
--- FAIL: TestBasicWorkflow (126.82s)
    --- PASS: TestBasicWorkflow/returns_no_service_before_running_tests (0.05s)
    --- PASS: TestBasicWorkflow/create_hello_service_and_returns_no_error (24.58s)
    --- PASS: TestBasicWorkflow/create_hello_service_again_and_get_service_already_exists_error (0.12s)
    --- PASS: TestBasicWorkflow/returns_valid_info_about_hello_service (0.14s)
    --- PASS: TestBasicWorkflow/update_hello_service's_configuration_and_returns_no_error (0.11s)
    --- PASS: TestBasicWorkflow/create_another_service_and_returns_no_error (2.99s)
    --- FAIL: TestBasicWorkflow/returns_a_list_of_revisions_associated_with_hello_and_svc2_services (0.13s)
        basic_workflow_test.go:132: assertion failed: 
            Actual output: hello-kxgvk   hello     3s    1 OK / 5     Unknown   Deploying
            Missing strings: True
...
```

Might be flaky and timing to allow status to update.

navidshaikh commented on test/e2e/basic_workflow_test.go:
```diff
@@ -132,6 +155,14 @@ func (test *e2eTest) serviceUpdate(t *testing.T, serviceName string, args []stri
+	out, err := test.kn.RunWithOpts([]string{"service", "describe", serviceName, "-o=name"}, runOpts{})
+	assert.NilError(t, err)
+
+	expectedName := fmt.Sprintf("service.serving.knative.dev/%s", serviceName)
```
this error message needs to be fixed on kn side, I will have a PR for this fix.

navidshaikh commented:
/test pull-knative-client-integration-tests-latest-release
